src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -501,9 +503,114 @@
 HYBRID_ALL_EXPENSIVE_METHODS = (
     HYBRID_HIGH_OVERHEAD_METHODS + HYBRID_ITERATIVE_STYLE_METHODS
 )
-# Set of (class name, method name) tuples for methods that are wholly unimplemented by
+
+
+# Named tuple for method registry keys.
+class MethodKey(NamedTuple):
+    api_cls_name: Optional[str]
+    method_name: str
+
+
+# Rule defining which args should trigger auto-switching.
+@dataclass
+class UnsupportedArgsRule:
+    """
+    Rule for defining argument combinations that trigger auto-switching to pandas.
+
+    Attributes:
+        unsupported_conditions: List of conditions that can be either:
+            - tuple[Callable, str]: (condition_function, reason) for complex conditions
+            - tuple[str, Any]: (argument_name, unsupported_value) for simple value checks
+    """
+
+    unsupported_conditions: List[Union[Tuple[Callable, str], Tuple[str, Any]]] = field(
+        default_factory=list
+    )
+
+
+# Set of MethodKey objects for methods that are wholly unimplemented by
 # Snowpark pandas. This list is populated by the register_*_not_implemented decorators.
-HYBRID_SWITCH_FOR_UNIMPLEMENTED_METHODS: Set[Tuple[str, str]] = set()
+HYBRID_SWITCH_FOR_UNIMPLEMENTED_METHODS: Set[MethodKey] = set()
+
+# Global registry for args-based switching rules
+HYBRID_SWITCH_FOR_UNSUPPORTED_ARGS: dict[MethodKey, UnsupportedArgsRule] = {}
+
+
+def register_query_compiler_method_not_implemented(
+    api_cls_name: Optional[str],
+    method_name: str,
+    unsupported_args: Optional["UnsupportedArgsRule"] = None,
+) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+    """
+    Decorator for SnowflakeQueryCompiler methods with args-based auto-switching.
+
+    Registers pre-op switching and raises a NotImplementedError if any unsupported-parameter predicate evaluates True.
```

Comment:
```suggestion
    Registers pre-op switching for the specified API-layer method, replacing the decorated query compiler method with a version that raises a NotImplementedError if any unsupported parameter predicate evaluates True.
```

It would also be helpful to add a bit in this docstring explaining the design reasoning for why the annotation is on the query compiler instead of on frontend (DataFrame/Series) methods.

Reply:
yup, I added more explanation on why doing it in the qc